File: src/Framework/Hadouken.Framework/Rpc/JsonRpcClient.cs
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Serialization;
using System;

namespace Hadouken.Framework.Rpc
{
    public sealed class JsonRpcClient : IJsonRpcClient
    {
        private readonly IClientTransport _transport;
        private static readonly JsonSerializerSettings SerializerSettings = new JsonSerializerSettings();
        private int _requestId = 0;

        static JsonRpcClient()
        {
            SerializerSettings.ContractResolver = new CamelCasePropertyNamesContractResolver();
            SerializerSettings.Converters.Add(new SemanticVersionConverter());
            SerializerSettings.Converters.Add(new StringEnumConverter());
            SerializerSettings.Converters.Add(new VersionConverter());
        }

        public JsonRpcClient(IClientTransport transport)
        {
            _transport = transport;
        }

        public TResult Call<TResult>(string method, object parameters = null)
        {
            _requestId++;

            var request = new
                {
                    id = _requestId,
                    jsonrpc = "2.0",
                    method,
                    @params = parameters
                };

            var json = JsonConvert.SerializeObject(request, SerializerSettings);
            var response = _transport.Send(json);

            JsonRpcResponse rpcResponse;
            Exception exception;

            if (JsonRpcResponse.TryParse(response, out rpcResponse, out exception))
            {
                var successResponse = rpcResponse as JsonRpcSuccessResponse;

                if (successResponse != null && successResponse.Result != null)
                {
                    return successResponse.Result.ToObject<TResult>();
                }
            }

            return default(TResult);
        }

        public void Dispose()
        {
        }
    }
}

File: src/Plugins/Web/Hadouken.Plugins.Web/UI/Scripts/Framework/Http/JsonRpcClient.ts
module Hadouken.Http {
    export class JsonRpcClient {
        private _url: string;
        private _requestId: number = 0;

        constructor(url: string) {
            this._url = url;
        }

        call(method: string, callback: { (result: any): void; }): void {
            this.callParams(method, null, callback);
        }

        callParams(method: string, params: any, callback: { (result: any): void; }): void {
            this._requestId += 1;

            var data = {
                id: this._requestId,
                jsonrpc: "2.0",
                method: method,
                params: params
            };

            var jsonData = JSON.stringify(data);

            $.post(this._url, jsonData, (response: any) => {
                if (typeof response.result === "undefined")
                    return;

                callback(response.result);
            });
        }
    }
}